Add support for multi-register circuit conversion

gcalin opened the pull request:
# Description

Currently, the `tk_to_qulacs` function converts the indices of the qubits upon gates operate in an _absolute_ manner, i.e., retrieve the qubit's index from its register. For example, for two qubit gates:

```python
  id1 = index_map[com.qubits[0].index[0]]
  id2 = index_map[com.qubits[1].index[0]]
```

As a result, circuits with multiple registers are not converted correctly. The corresponding Qulacs circuit that the function builds only has a single register, and `index_map` does not account for the register's position in the original tket circuit. I added two functions that build a dictionary in such a way that multiple registers are accounted for in the input circuit.

yao-cqc commented:
@gcalin Hi, thanks for the PR! Could you provide an example that produces the problem? We do have a default `FlattenRegisters` pass in the compilation stack that should merge all registers into a single register.

gcalin commented:
@yao-cqc Thank you for looking into this. I am encountering this error when using a different compilation stack. Particularly, the workflow I am using is (1) building a circuit in `qiskit` -> (2) transpiling in `qiskit` (to a compatible gate set) -> (3) converting (directly, no compilation) to `qulacs` using `pytket-qiskit` and `pytket-qulacs`.

I am new to `tket` and I understand the usual workflow includes a (`tket`) compilation step (and the flattening pass) - but is that strictly necessary to use the `tk_to_qulacs` conversion function?

yao-cqc commented:
> @yao-cqc Thank you for looking into this. I am encountering this error when using a different compilation stack. Particularly, the workflow I am using is (1) building a circuit in `qiskit` -> (2) transpiling in `qiskit` (to a compatible gate set) -> (3) converting (directly, no compilation) to `qulacs` using `pytket-qiskit` and `pytket-qulacs`.
> 
> I am new to `tket` and I understand the usual workflow includes a (`tket`) compilation step (and the flattening pass) - but is that strictly necessary to use the `tk_to_qulacs` conversion function?

Got you. The documentation for `tk_to_qulacs` is a bit lacking atm. It assumes that the input circuit only has the default quantum register. You can try run the [level 0](https://github.com/CQCL/pytket-qulacs/blob/e17574ad603e03e1d6f1739779bafbc5617c676b/pytket/extensions/qulacs/backends/qulacs_backend.py#L161) compilation pass, which doesn't do any optimisation and only does box decomposition, rebase (if the gatesets don't match) and register merging. 

If you don't like the way how FlattenRegisters behaves, you also have the option to use `Circuit.rename_units` to manually merge your registers. [API doc](https://cqcl.github.io/tket/pytket/api/circuit_class.html#pytket.circuit.Circuit.rename_units), [example](https://github.com/CQCL/tket/blob/4e62011109ca5669ecebb335abbbf9efa1e6367a/pytket/tests/predicates_test.py#L877)

yao-cqc commented:
For example:

```python
from pytket.extensions.qulacs.qulacs_convert import (
    tk_to_qulacs,
)
from pytket.extensions.qulacs.backends import QulacsBackend
from pytket.circuit import Circuit, Qubit

c = Circuit()
reg_A = c.add_q_register("A", 1)
reg_B = c.add_q_register("B", 2)
c.CX(reg_A[0], reg_B[1])
# option 1: use the default compilation pass
b = QulacsBackend()
d = b.get_compiled_circuit(c, optimisation_level=0)
qulacs_circ = tk_to_qulacs(d)

# option 2: manually rename
c.rename_units({
    reg_A[0]: Qubit(0),
    reg_B[0]: Qubit(1),
    reg_B[1]: Qubit(2),
})
qulacs_circ = tk_to_qulacs(c)
```

of course you can also call FlattenRegisters directly
```python
from pytket.passes import FlattenRegisters
FlattenRegisters().apply(c)
```

gcalin commented:
> > @yao-cqc Thank you for looking into this. I am encountering this error when using a different compilation stack. Particularly, the workflow I am using is (1) building a circuit in `qiskit` -> (2) transpiling in `qiskit` (to a compatible gate set) -> (3) converting (directly, no compilation) to `qulacs` using `pytket-qiskit` and `pytket-qulacs`.
> > I am new to `tket` and I understand the usual workflow includes a (`tket`) compilation step (and the flattening pass) - but is that strictly necessary to use the `tk_to_qulacs` conversion function?
> 
> Got you. The documentation for `tk_to_qulacs` is a bit lacking atm. It assumes that the input circuit only has the default quantum register. You can try run the [level 0](https://github.com/CQCL/pytket-qulacs/blob/e17574ad603e03e1d6f1739779bafbc5617c676b/pytket/extensions/qulacs/backends/qulacs_backend.py#L161) compilation pass, which doesn't do any optimisation and only does box decomposition, rebase (if the gatesets don't match) and register merging.
> 
> If you don't like the way how FlattenRegisters behaves, you also have the option to use `Circuit.rename_units` to manually merge your registers. [API doc](https://cqcl.github.io/tket/pytket/api/circuit_class.html#pytket.circuit.Circuit.rename_units), [example](https://github.com/CQCL/tket/blob/4e62011109ca5669ecebb335abbbf9efa1e6367a/pytket/tests/predicates_test.py#L877)

Understood, thank very much you for the references and examples. In this case, would you be interested in extending the functionality of `tk_to_qulacs` to arbitrary circuits (i.e., by performing checks on presupposed properties, like there being a single register, which means the conversion might perform compilation subroutines?), or is the "basic" (`optimisation_level=0`) compilation a must for the input circuit?

yao-cqc commented:
> > > @yao-cqc Thank you for looking into this. I am encountering this error when using a different compilation stack. Particularly, the workflow I am using is (1) building a circuit in `qiskit` -> (2) transpiling in `qiskit` (to a compatible gate set) -> (3) converting (directly, no compilation) to `qulacs` using `pytket-qiskit` and `pytket-qulacs`.
> > > I am new to `tket` and I understand the usual workflow includes a (`tket`) compilation step (and the flattening pass) - but is that strictly necessary to use the `tk_to_qulacs` conversion function?
> > 
> > 
> > Got you. The documentation for `tk_to_qulacs` is a bit lacking atm. It assumes that the input circuit only has the default quantum register. You can try run the [level 0](https://github.com/CQCL/pytket-qulacs/blob/e17574ad603e03e1d6f1739779bafbc5617c676b/pytket/extensions/qulacs/backends/qulacs_backend.py#L161) compilation pass, which doesn't do any optimisation and only does box decomposition, rebase (if the gatesets don't match) and register merging.
> > If you don't like the way how FlattenRegisters behaves, you also have the option to use `Circuit.rename_units` to manually merge your registers. [API doc](https://cqcl.github.io/tket/pytket/api/circuit_class.html#pytket.circuit.Circuit.rename_units), [example](https://github.com/CQCL/tket/blob/4e62011109ca5669ecebb335abbbf9efa1e6367a/pytket/tests/predicates_test.py#L877)
> 
> Understood, thank very much you for the references and examples. In this case, would you be interested in extending the functionality of `tk_to_qulacs` to arbitrary circuits (i.e., by performing checks on presupposed properties, like there being a single register, which means the conversion might perform compilation subroutines?), or is the "basic" (`optimisation_level=0`) compilation a must for the input circuit?

To make `tk_to_qulacs` more useful as a standalone function, I think it should be suffice to add a check for `circuit.is_simple()` ([doc](https://cqcl.github.io/tket/pytket/api/circuit_class.html#pytket.circuit.Circuit.is_simple)). Other properties are already being checked by this [line](https://github.com/CQCL/pytket-qulacs/blob/e17574ad603e03e1d6f1739779bafbc5617c676b/pytket/extensions/qulacs/qulacs_convert.py#L99).

gcalin commented:
> > > > @yao-cqc Thank you for looking into this. I am encountering this error when using a different compilation stack. Particularly, the workflow I am using is (1) building a circuit in `qiskit` -> (2) transpiling in `qiskit` (to a compatible gate set) -> (3) converting (directly, no compilation) to `qulacs` using `pytket-qiskit` and `pytket-qulacs`.
> > > > I am new to `tket` and I understand the usual workflow includes a (`tket`) compilation step (and the flattening pass) - but is that strictly necessary to use the `tk_to_qulacs` conversion function?
> > > 
> > > 
> > > Got you. The documentation for `tk_to_qulacs` is a bit lacking atm. It assumes that the input circuit only has the default quantum register. You can try run the [level 0](https://github.com/CQCL/pytket-qulacs/blob/e17574ad603e03e1d6f1739779bafbc5617c676b/pytket/extensions/qulacs/backends/qulacs_backend.py#L161) compilation pass, which doesn't do any optimisation and only does box decomposition, rebase (if the gatesets don't match) and register merging.
> > > If you don't like the way how FlattenRegisters behaves, you also have the option to use `Circuit.rename_units` to manually merge your registers. [API doc](https://cqcl.github.io/tket/pytket/api/circuit_class.html#pytket.circuit.Circuit.rename_units), [example](https://github.com/CQCL/tket/blob/4e62011109ca5669ecebb335abbbf9efa1e6367a/pytket/tests/predicates_test.py#L877)
> > 
> > 
> > Understood, thank very much you for the references and examples. In this case, would you be interested in extending the functionality of `tk_to_qulacs` to arbitrary circuits (i.e., by performing checks on presupposed properties, like there being a single register, which means the conversion might perform compilation subroutines?), or is the "basic" (`optimisation_level=0`) compilation a must for the input circuit?
> 
> To make `tk_to_qulacs` more useful as a standalone function, I think it should be suffice to add a check for `circuit.is_simple()` ([doc](https://cqcl.github.io/tket/pytket/api/circuit_class.html#pytket.circuit.Circuit.is_simple)). Other properties are already being checked by this [line](https://github.com/CQCL/pytket-qulacs/blob/e17574ad603e03e1d6f1739779bafbc5617c676b/pytket/extensions/qulacs/qulacs_convert.py#L99).

I removed my initial solution and replaced it with a conditional `FlattenRegisters` pass. This indeed makes it possible to circumvent the rest of the compilation stack for my use case.

yao-cqc reviewed (approved):
Looks good, thank you @gcalin !
